app/controllers/api/v1/questionnaires_controller.rb

```diff
@@ -0,0 +1,169 @@
+        (current_user_is_a?('Teaching Assistant') && session[:user].instructor_id == @questionnaire.try(:instructor_id))
+    else
+      current_user_has_student_privileges?
+    end
+  end
+
+  #List all the questionnaires
+  # GET on /questionnaires
+  def index
+    @questionnaires = Questionnaire.order(:id)
+    render json: @questionnaires
+  end
+  
+  #Show a given questionnaire
+  # GET on /questionnaires/:id
+  def show
+    begin
+      @questionnaire = Questionnaire.find(params[:id])
+      render json: @questionnaire
+    rescue
+      msg = "No such Questionnaire exists."
+      render json: msg, status: :not_found
+    end
+  end
+  
+  #Create a given questionnaire
+  # POST on /questionnaires
+  def create
+    if params[:questionnaire][:name].blank?
+      redirect_to controller: 'questionnaires', action: 'new', model: params[:questionnaire][:type], private: params[:questionnaire][:private]
+    else
+      questionnaire_private = params[:questionnaire][:private] == 'true'
+      display_type = params[:questionnaire][:type].split('Questionnaire')[0]
+      begin
+        @questionnaire = Object.const_get(params[:questionnaire][:type]).new if Questionnaire::QUESTIONNAIRE_TYPES.include? params[:questionnaire][:type]
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+      begin
+        @questionnaire.private = questionnaire_private
+        @questionnaire.name = params[:questionnaire][:name]
+        @questionnaire.instructor_id = 6 # session[:user].id
+        @questionnaire.min_question_score = params[:questionnaire][:min_question_score]
+        @questionnaire.max_question_score = params[:questionnaire][:max_question_score]
+        @questionnaire.type = params[:questionnaire][:type]
+        if %w[AuthorFeedback CourseSurvey TeammateReview GlobalSurvey AssignmentSurvey BookmarkRating].include?(display_type)
+          display_type = (display_type.split(/(?=[A-Z])/)).join('%')
+        end
+        @questionnaire.display_type = display_type
+        @questionnaire.instruction_loc = Questionnaire::DEFAULT_QUESTIONNAIRE_URL
+        @questionnaire.save
+        # Create node
+        tree_folder = TreeFolder.where(['name like ?', @questionnaire.display_type]).first
+        parent = FolderNode.find_by(node_object_id: tree_folder.id)
+        QuestionnaireNode.create(parent_id: parent.id, node_object_id: @questionnaire.id, type: 'QuestionnaireNode')
+        render json: @questionnaire, status: :created
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+     
+    end
+  end
+
+  # Remove a given questionnaire
+  # DELETE on /questionnaires/:id
+  def destroy
+    begin
+      @questionnaire = Questionnaire.find(params[:id])
+    rescue
+      render json: $ERROR_INFO
+    end
+    if @questionnaire
+      begin
+        name = @questionnaire.name
+        questions = @questionnaire.questions
+        questions.each do |question|
```

Comment:
we don't need to delete questions here if there are questions associated with the questionnaire, the controller will respond with an appropriate error something like - can't delete since there are questions and provide a link with which we can delete all questions that belong to this questionnaire

Reply:
Commit 285c94a resolves this review